Beatless initial PR

ph opened the pull request:
## This PR provide the following

1. Plugin infra for developing providers and functions
2. A local stdin provider only used for testing, I will remove it in the
final version.
3. AWS provider and function types for:
  - Cloudwatch logs
  - SQS
  - Kinesis
  - Api web gateway proxy
4. License checker
5. Packaging of artifact
6. Runners
7. CLI infrastructure
8. CLI to push a cloudwatch logs function.
9. CLI to delete any function
10. Processors support.
11. Types to validate value from the users and the lambda function.

## What it doesn't provides

- ECS and full event extraction. (NOT for v1)
- Specifying the AWS credentials in the configuration
- CLI for SQS, Kinesis, API
- Robust CLI interaction with the API, rollback on failure / versioning.
- Removal of not supported outputs
- Removal of seccomp check
- Integration tests
- Updated build task to produce containing the user executable beat and
the linux beats.
- Concurrency / memory settings

## Vendored

- aws/aws-lambda-go
- aws/aws-sdk-go-v2

---

**Note:** The PR look much bigger because of the vendored files.

**Notes 2:** 

Make sure you add this to your config YAMl, I wan't to make the following the default options but it would have been really messy to do it in this PR.

```
# Disable anything that could be send to disk
path.data: /tmp
path.logs: /tmp/logs
logging.to_stderr: true
logging.to_files: false
#logging.level: debug
setup.template.enabled: true
queue.mem:
  events: 50
  flush.min_events: 1
  flush.timeout: 0.5s
```

## related work
- https://github.com/elastic/beats/pull/8449
- https://github.com/elastic/beats/pull/8443

ph commented:
@kvch @ruflin go ahead for first round of review, I am still working on the CLI so expect them to changes but I see no reason to not start reviewing this.

ph commented:
I think I will migrate the to the cloud formation api instead this look like a more solid way of doing it.

ph commented:
@urso @ruflin @kvch Do not takes too much times into reviewing the CLI code in the cloudwatchlogs.go I've migrated everything to the cloudformation API which provided a lot more stable experience.

kvch commented on x-pack/beatless/provider/local/local.go:
```diff
@@ -0,0 +1,101 @@
+	"github.com/elastic/beats/x-pack/beatless/provider"
+)
+
+var stdinName = "stdin"
```
`const`?

kvch commented on x-pack/beatless/provider/local/local.go:
```diff
@@ -0,0 +1,101 @@
+		scanner.Split(bufio.ScanLines)
+
+		for scanner.Scan() {
+			if scanner.Err() != nil {
```
It's enough to call `scanner.Err` once, if you rewrite the code a little:
```golang
if err := scanner.Err(); err != nil {
    errChan <- err
    return
}
```

kvch commented on x-pack/beatless/provider/aws/executor.go:
```diff
@@ -0,0 +1,113 @@
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+var errNeverRun = errors.New("executer was never executed")
```
```golang
var (
    errNeverRun = errors.New("executer was never executed")
    errCannotAdd = errors.New("cannot add to an already executed executer")
    errAlreadyExecuted = errors.New("executer already executed")
)
```

kvch commented on x-pack/beatless/provider/aws/executor.go:
```diff
@@ -0,0 +1,113 @@
+
+func newExecutor(log *logp.Logger, context *executorContext) *executor {
+	if log == nil {
+		log = logp.NewLogger("executer")
```
This probably should be `"executor"`.

kvch commented on x-pack/beatless/provider/aws/executor.go:
```diff
@@ -0,0 +1,113 @@
+		}
+	}
+	if err == nil {
+		e.log.Debug("all operations successful")
```
`"all operations have been successful"`?

ph commented:
@ruflin As discussed I have added the following command `beatless function package` this will generate a package that you can push manually or via another CLI to lambda. By default it will create a _package.zip_ in the working folder, but you can override that with `-o`

Also, this PR is a lot more robust when deploying, updating or removing a cloudwatch function, it require permission on cloudformation, but it will make the operation atomic, and will take into consideration the changes that you add.

I have updated the configuration to correctly expose lambda specific settings:

```yaml
beatless.provider.aws.functions:
  # Define the list of function availables, each function required to have a unique name.
  - name: fn_cloudwatch_logs
    type: cloudwatch_logs

    # Description of the method to help identify them when you run multiples functions.
    description: "listen to cloudwatch logs"
    #
    # Concurrency, is the reserved number of instance for that function.
    # Default is unreserved.
    #
    # Note: There is a hard limit of 1000 of running function of any kind per account.
    # concurrency: 5
    #
    # The maximum memory allocated for this function, the configured size must be a factor of 64.
    # There is a hard limit of 3008MiB for each function. Default is 128MiB.
    # memory_size: 128MiB
    #
    # Dead letter queue configuration, this must be set to an ARN pointing to a SQS queue.
    # dead_letter_config.target_arn:
    #
    # Optional fields that you can specify to add additional information to the
    # output. Fields can be scalar values, arrays, dictionaries, or any nested
    # combination of these.
    #fields:
    #  env: staging

    # List of cloudwatch log group registered to that function.
    triggers:
      - log_group_name: /aws/lambda/beatless-cloudwatch_logs
        filter_pattern: mylog_
    #
    # Define custom processors for this function.
    # processors:
    #   - dissect:
    #       tokenizer: "%{key1} %{key2}"

```
Once the PR #8573 is merged it will be a PR the same size that you had before.

ph commented:
@ruflin @kvch @urso All rebase and ready to get looked at. :)

ph commented on x-pack/beatless/provider/aws/cloudwatch_logs.go:
```diff
@@ -0,0 +1,179 @@
+
+// AWSLogsSubscriptionFilter overrides the type from goformation to allow to pass an empty string.
+// The API support an empty string, but requires one, the original type does not permit that.
+type AWSLogsSubscriptionFilter struct {
```
For clarity I do not agree with Hound here.

ph commented on x-pack/beatless/provider/aws/cli_manager.go:
```diff
@@ -0,0 +1,328 @@
+
+// AWSLambdaFunction add 'dependsOn' as a serializable parameters, for no good reason it's
+// not supported.
+type AWSLambdaFunction struct {
```
For clarity I do not agree with Hound here.

ph commented:
OK I have SQS and APIGateway working with the cloudformation API, I will do a followup PR with them since I use the same strategy as the cloudwatchlog API.

urso commented on x-pack/beatless/provider/default_provider.go:
```diff
@@ -0,0 +1,86 @@
+}
+
+// NewDefaultProvider returns factory methods to handle generic provider.
+func NewDefaultProvider(name string, manager CLIManagerFactory) func(*logp.Logger, *Registry, *common.Config) (Provider, error) {
```
Do we have a type for the function return type?

urso commented on x-pack/beatless/provider/default_provider.go:
```diff
@@ -0,0 +1,86 @@
+// NewNullCli returns a NOOP CliManager.
+func NewNullCli(_ *logp.Logger, _ *common.Config, _ Provider) (CLIManager, error) {
+	return &NullCLI{}, nil
+}
```
common pattern is to use a singleton type:

```
type nullCLI struct{}

func NewNullCli(...) (CLIManager, error) {
  return (*nullCLI)(nil), nil
}

func (_ *nullCLI) Deploy(_ string) error { ... }
func (_ *nullCLI) Update(_ string) error { ... }
func (_ *nullCLI) Remove(_ string) error { ... }
```

This does not allocate. An interface value is the product of type + value. The trick is to assign the type `(*nullCLI)` and the value `nil` -> The interface value is not `nil` and has an executable vtable.

ph commented on libbeat/feature/registry.go:
```diff
@@ -28,26 +28,26 @@ import (
 // feature are grouped by namespace, a namespace is a kind of plugin like outputs, inputs, or queue.
 // The feature name must be unique.
-type registry struct {
+type FeatureRegistry struct {
```
Yeah much better, it will lead to less fragile code. :)

ph commented on x-pack/beatless/.gitignore:
```diff
@@ -3,7 +3,9 @@
 data
 fields.yml
 logs
+_meta/deploy
```
This is some left over and we should use user defined ignore file instead, I've removed the change.

ph commented on dev-tools/generate_notice.py:
```diff
@@ -67,6 +68,7 @@ def gather_dependencies(vendor_dirs, overrides=None):
                 lib["license_contents"] = read_file(lib["license_file"])
                 lib["license_summary"] = detect_license_summary(lib["license_contents"])
                 if lib["license_summary"] == "UNKNOWN":
+                    print("no valid: {}".format(lib["license_file"]))
```
Its a left over, its removed.

ph commented on x-pack/beatless/_meta/beat.reference.yml:
```diff
@@ -0,0 +1,49 @@
+    #
+    # The maximum memory allocated for this function, the configured size must be a factor of 64.
+    # There is a hard limit of 3008MiB for each function. Default is 128MiB.
+    # memory_size: 128MiB
```
There is validation in place for this value.

urso replied:
WFM

ph commented on x-pack/beatless/config/config.go:
```diff
@@ -2,11 +2,73 @@
+		"flush.min_events": 1,
+		"flush.timeout":    "0.1s",
+	},
+})
```
Yup, super hacky, but I haven't found a good way without that :(
I would have preferred to be able to define internal config type instead of providing an untyped common.Config. I expected the reference syntax to work.

ph commented on x-pack/beatless/config/config.go:
```diff
@@ -2,11 +2,73 @@
+
 // Config default configuration for Beatless.
 type Config struct {
+	Provider *common.ConfigNamespace `config:"provider" validate:"required"`
```
This is exact, when I've made the initial version I have actually taken the `map[string]*common.Config` way of defining providers, once I've started playing more with providers and looking at the configuration of providers, I've decided to make them unique per config. I thought it was clearer as a user to only be able to define a unique provider per functionbeat instance.

I think we can revisit that later once we got more than one provider or we have gathered user feedback.

urso replied:
SGTM

ph commented on x-pack/beatless/config/config.go:
```diff
@@ -2,11 +2,73 @@
+
+		if _, found := names[fc.Name]; found {
+			return fmt.Errorf("function name '%s' already exist, name must be unique", fc.Name)
+		}
```
I wanted to be close to the way we define processors but I would be able to change it to key/value if you think its better?

urso replied:
No strong feelings in any directions. I'm fine with list syntax, we already do this in most places anyways.

ph commented on x-pack/beatless/cmd/cli_handler.go:
```diff
@@ -0,0 +1,131 @@
+		return fmt.Errorf("fail to deploy %d function(s)", errCount)
+	}
+	return nil
+}
```
> Have you considered to run the operations in parallel?

I did consider it but I've decided against it to limit any concurrency issue I could get on the provider, but we could make the provider CLI implement a parallel or a serial interface and we could let the cli handler apply the right execution model. I won't do it for v1.

ph replied:
I've used `fmt.Fprintf()` instead of a logger because theses are user visible error message like any command line. So there are not really log statements.

ph commented on x-pack/beatless/cmd/cli_handler_test.go:
```diff
@@ -0,0 +1,136 @@
+		m.On("Remove", "super").Return(nil)
+		m.On("Remove", "ninja").Return(nil)
+		handler := newCLIHandler(m, errOut, output)
+		err := handler.Remove([]string{"super", "ninja"})
```
its over 9000

ph commented on x-pack/beatless/cmd/provider_cmd.go:
```diff
@@ -0,0 +1,124 @@
+		}),
+	}
+	return cmd
+}
```
My idea of clihandler was to separate the burden of multiple functions vs  a single function, maybe that will change once we have more than one provider. I would keep it as is with you suggested refactoring and reevaluate if we still need it once we have gcq or azure function support.

ph commented on x-pack/beatless/core/bundle/bundle.go:
```diff
@@ -0,0 +1,206 @@
+// NoOpCloser wrap an io.Reader into a io.ReadClose and make the close operation do nothing.
+type NoOpCloser struct {
+	reader io.Reader
+}
```
I didn't even know that this actually exists. 🤦‍♂️

ph commented on x-pack/beatless/core/sync_client.go:
```diff
@@ -83,13 +85,15 @@ func NewSyncClient(pipeline beat.Pipeline, cfg beat.ClientConfig) (*SyncClient,
 
 // PublishAll publish a slice of events to the pipeline and return.
 func (s *SyncClient) PublishAll(events []beat.Event) error {
+	s.log.Debugf("publish all, events count: %d", len(events))
```
@urso yes logger sound wonderful, our current tool makes it really clunky to have readable messages but still make it easy for machine parsing.

ph commented on x-pack/beatless/licenser/elastic_fetcher.go:
```diff
@@ -82,15 +84,25 @@ func (et *expiryTime) UnmarshalJSON(b []byte) error {
+		path string,
+		pipeline string,
+		params map[string]string,
+		body interface{},
```
I wanted to be able to use the elasticsearch client directly with the fetch or wrap it in the mux without changing the original code.

ph commented on x-pack/beatless/beater/beatless.go:
```diff
@@ -5,68 +5,157 @@
+		// Make the client aware of the current license, the client will accept sending events to the
+		// pipeline until the client is closed or if the license change and is not valid.
+		licenseAware := core.NewLicenseAwareClient(client, checkLicense)
+		manager.AddWatcher(licenseAware)
```
I have added the check, it should only happen when a watcher is added twice.

ph commented on x-pack/beatless/licenser/manager.go:
```diff
@@ -309,3 +312,32 @@ func (m *Manager) invalidate() {
+
+	if err := manager.AddWatcher(&callback); err != nil {
+		return err
+	}
```
Not it cannot, it just a check agains't the structure that keep track of the watcher.

ph commented on x-pack/beatless/core/makezip.go:
```diff
@@ -0,0 +1,58 @@
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package core
```
It was more mean to provide a core framework, I am OK to split that up in another PR.

ph commented on x-pack/beatless/Makefile:
```diff
@@ -14,3 +14,9 @@ include $(ES_BEATS)/libbeat/scripts/Makefile
 collect:
+
+# TODO(ph) This is used for debugging until we change the build to create 2 artifacts,
+# we will do this in another PR.
```
I will create a followup PR.

ph commented on x-pack/beatless/core/bundle/bundle_test.go:
```diff
@@ -0,0 +1,99 @@
+		panic(err)
+	}
+	return &MemoryFile{Path: name, Raw: paddingData, FileMode: 0755}
+}
```
I have added physical test for that. 👍

ph commented on x-pack/beatless/provider/aws/api_gateway_proxy.go:
```diff
@@ -0,0 +1,84 @@
+			request.RequestContext.RequestID,
+		), nil
+	}
+}
```
By making it a method it make it really simple to test any interaction without having to deal with the blocking _lambda.Start_.

urso replied:
What about:

```
type apiHandler func(request events.APIGatewayProxyRequest) (events.APIGatewayProxyResponse, error)

func createHandler(log *logp.Logger, client core.Client) apiHandler {
  return withResponse(func(request events.APIGatewayProxyRequest) error {
    event := transformer.APIGatewayProxyRequest(request)
    if err := client.Publish(event); err != nil {
      log.Errorf(...)
      return err
    }

    client.Wait()
    return nil
  })
}

func withResponse(fn func(event.APIGatewayProxyRequest) error) apiHandler {
  return func(request events.APIGatewayProxyRequest) (events.APIGatewayProxyResponse, error) {
    err := fn(request)
    if err != nil {
      return buildResponse(http.StatusInternalServerError,
				"an error occurred when sending the event.",
				request.RequestContext.RequestID), err
    }
    
    return buildResponse(
			http.StatusOK,
			"event received successfully.",
			request.RequestContext.RequestID,
	), nil
  }
}
```

I wonder a little about best practices in regards to:
- error handling (we return an error value, but also create error message in response)
- common way of error reporting in the response (e.g. how/when to construct from errors)
- patterns for creating and maintaining handlers

It's somewhat similar to plain HTTP server in the stdlib, yet people follow some different/same patterns.

ph replied:
@urso let me think about it a bit :)

ph commented on x-pack/beatless/provider/aws/api_gateway_proxy.go:
```diff
@@ -0,0 +1,84 @@
+		Headers:    map[string]string{"Content-Type": "application/json"},
+		Body:       string(body),
+	}
+}
```
I think so, there is two level of response in this code, the actual _APIGatewayProxyResponse_ which is what the AWS API will look but also the actual JSON response for the remote caller. Using a method allow to make sure that happy and non happy path have the same behavior.

urso replied:
I was wondering if a function will do. I don't see any field of `a` being used.

ph commented on x-pack/beatless/provider/aws/aws.go:
```diff
@@ -0,0 +1,41 @@
+	"aws",
+	provider.NewDefaultProvider("aws", NewCLI),
+	feature.NewDetails("AWS Lambda", "listen to events on AWS lambda", feature.Experimental),
+).MustAddFunction("cloudwatch_logs",
```
I've wrapped the feature API inside a builder pattern, because the way we have designed the registry is to have a flat view of the namespace keys, using a builder pattern make sure that the feature linked to an existing provider are correctly persisted in the right namespaces. This reduce risk of name errors, the builder is just a sugar implementation over the `feature.Bundle`, this make sure that even sub feature use the same global registry.

ph commented on x-pack/beatless/provider/aws/aws.go:
```diff
@@ -0,0 +1,41 @@
+		"Cloudwatch Logs trigger",
+		"receive events from cloudwatch logs.",
+		feature.Experimental,
+	)).MustAddFunction("api_gateway_proxy",
```
I did the change, my vim formatter doesn't like it :)

ph commented on x-pack/beatless/provider/aws/cli_manager.go:
```diff
@@ -0,0 +1,328 @@
+
+	prefix := func(s string) string {
+		return "btl" + name + s
+	}
```
Created a followup issue at https://github.com/elastic/beats/issues/8635

ph replied:
Created a followup issue at https://github.com/elastic/beats/issues/8635

ph commented on x-pack/beatless/provider/aws/executor.go:
```diff
@@ -0,0 +1,99 @@
+	if err == nil {
+		e.log.Debug("rollback successful")
+	} else {
+		e.log.Debug("rollback incomplete")
```
The execution was a bit more involved before I have switched to cloudformation.

If we take a deployment workflow of a cloudwatch function, the following will happen:

- We generate the zip
- we push the zip to the bucket (can fail, nothing is done but we could delete the artifact)
- We generate the cloudformation template (can fail, we don't need to rollback)
- We push the cloudformation template to the bucket (can fail, nothing is done but we could delete the file.
- We send a _Create_ or _Update_ stack request to cloudformation, at this point the rollback is in AWS hand nothing is required on the beats side.
- We wait for the stack to converge.

I've keep the rollback logic to make sure we could remove stuff in future iteration. 
I don't think we need to catch a panic here a panic would mean that we have something really bad in the API SDK and I am not sure we could recover.

urso replied:
> I've keep the rollback logic to make sure we could remove stuff in future iteration.

👍 

> I don't think we need to catch a panic here a panic would mean that we have something really bad in the API SDK and I am not sure we could recover.

👍 

+1 on keeping the undo logic. I was wondering if we would like to serialize the operations+state+reults to a file, such that we can continue or undo the operations on restart.

ph replied:
@urso Oh certainly could do that, maybe with the addition of others providers and having a lot of function to manage I would see the need for that.

ph commented:
@urso updated and fixed any formatting issue.

urso commented on x-pack/beatless/_meta/beat.reference.yml:
```diff
@@ -0,0 +1,49 @@
+    #
+    # Note: There is a hard limit of 1000 functions of any kind per account.
+    #concurrency: 5
+    #
```
s/#//

urso commented on x-pack/beatless/_meta/beat.reference.yml:
```diff
@@ -0,0 +1,49 @@
+    # The maximum memory allocated for this function, the configured size must be a factor of 64.
+    # There is a hard limit of 3008MiB for each function. Default is 128MiB.
+    #memory_size: 128MiB
+    #
```
s/#//

urso commented on x-pack/beatless/_meta/beat.reference.yml:
```diff
@@ -0,0 +1,49 @@
+    #
+    # Dead letter queue configuration, this must be set to an ARN pointing to a SQS queue.
+    # dead_letter_config.target_arn:
+    #
```
s/#//

urso commented on x-pack/beatless/_meta/beat.reference.yml:
```diff
@@ -0,0 +1,49 @@
+    #memory_size: 128MiB
+    #
+    # Dead letter queue configuration, this must be set to an ARN pointing to a SQS queue.
+    # dead_letter_config.target_arn:
```
s/# /#/

urso commented on x-pack/beatless/_meta/beat.reference.yml:
```diff
@@ -0,0 +1,49 @@
+    triggers:
+      - log_group_name: /aws/lambda/beatless-cloudwatch_logs
+        filter_pattern: mylog_
+    #
```
s/#//

urso commented on x-pack/beatless/_meta/beat.reference.yml:
```diff
@@ -0,0 +1,49 @@
+    # Define custom processors for this function.
+    # processors:
+    #   - dissect:
+    #       tokenizer: "%{key1} %{key2}"
```
s/# /#/

ph commented:
@urso updated with the changes.

urso commented on x-pack/beatless/beater/beatless.go:
```diff
@@ -5,68 +5,162 @@
+	// Each function has his own client to the publisher pipeline,
+	// publish operation will block the calling thread, when the method unwrap we have received the
+	// ACK for the batch.
+	clientFactory := func(cfg *common.Config) (core.Client, error) {
```
as makeClientFactory is the constructor now, just use `return func(cfg *common.Config) (core.Client, error)`. This makes it immediately clear that clientFactory will not be used inside this function.

urso commented on libbeat/feature/registry.go:
```diff
@@ -28,26 +28,26 @@ import (
 type mapper map[string]map[string]Featurable
 
-// Registry implements a global registry for any kind of feature in beats.
+// Registry implements a global FeatureRegistry for any kind of feature in beats.
```
The comment indeed reads somewhat funny.

urso commented on x-pack/beatless/provider/default_provider.go:
```diff
@@ -0,0 +1,86 @@
+}
+
+// Deploy returns not implemented.
+func (*nullCLI) Deploy(_ string) error { return fmt.Errorf("deploy not implemented") }
```
btw. If a type or the method is not exported, no godoc is required :)

In general +1 on having documentation on methods, but sometimes godoc style comments feel a little too shallow if they don't really add "value" in understanding the implementation.

urso commented on x-pack/beatless/core/coordinator.go:
```diff
@@ -0,0 +1,95 @@
+
+	err := rfn.Run(ctx)
+	if err != nil {
+		r.log.Error(
```
s/Error/Errorf/

urso commented on x-pack/beatless/provider/aws/api_gateway_proxy.go:
```diff
@@ -0,0 +1,84 @@
+		a.log.Debugf("received event (requestID: %s)", request.RequestContext.RequestID)
+		event := transformer.APIGatewayProxyRequest(request)
+		if err := client.Publish(event); err != nil {
+			a.log.Errorf("could not publish event to the pipeline, error: %s")
```
- missing parameter. 
- `s/%s/%+v/`

urso commented on x-pack/beatless/provider/aws/cli_manager.go:
```diff
@@ -0,0 +1,336 @@
+		return err
+	}
+
+	c.log.Debugf("cloudformation template: \n%s", json)
```
`s/: \n/:\n/`

urso commented on x-pack/beatless/provider/aws/cloudwatch_logs.go:
```diff
@@ -0,0 +1,183 @@
+
+		events := transformer.CloudwatchLogs(parsedEvent)
+		if err := client.PublishAll(events); err != nil {
+			c.log.Errorf("could not publish events to the pipeline, error: %s")
```
- missing parameter
- `s/%s/%+v`

urso commented on x-pack/beatless/provider/aws/kinesis.go:
```diff
@@ -0,0 +1,53 @@
+
+		events := transformer.KinesisEvent(request)
+		if err := client.PublishAll(events); err != nil {
+			k.log.Errorf("could not publish events to the pipeline, error: %s")
```
- missing parameter
- `s/%s/%+v`

ph commented:
Failures is not related to this PR.

ph commented:
I've rebased this PR on top of the feature branch you can now create artifact of beatless and push the function using theses artifacts.
